src/lib.rs

```diff
@@ -66,222 +36,460 @@ pub fn invoke(params: u32) -> u32 {
+    //     }
+    //
+    //     let ch_cid = checkpoint.cid();
+    //     // verify checkpoint
+    //     st.verify_checkpoint(&checkpoint)?;
+    //
+    //     // get votes for committed checkpoint
+    //     let mut votes_map = make_map_with_root::<_, Votes>(&st.window_checks, &Blockstore)
+    //         .map_err(|e| anyhow!("failed to load checkpoints: {}", e))?;
+    //     let mut found = false;
+    //     let mut votes = match get_votes(&votes_map, &ch_cid)? {
+    //         Some(v) => {
+    //             found = true;
+    //             v.clone()
+    //         }
+    //         None => Votes {
+    //             validators: Vec::new(),
+    //         },
+    //     };
+    //
+    //     if votes.validators.iter().any(|x| x == &caller) {
+    //         return Err(anyhow!("miner has already voted the checkpoint"));
+    //     }
+    //
+    //     // add miner vote
+    //     votes.validators.push(caller);
+    //
+    //     // if has majority
+    //     if st.has_majority_vote(&votes)? {
+    //         // commit checkpoint
+    //         st.flush_checkpoint::<&Blockstore>(&checkpoint)?;
+    //         // propagate to sca
+    //         st.send(
+    //             &Address::new_id(sca_actor_addr),
+    //             Method::CommitChildCheckpoint as u64,
+    //             RawBytes::serialize(checkpoint)?,
+    //             0.into(),
+    //         )?;
+    //         // remove votes used for commitment
+    //         if found {
+    //             votes_map.delete(&ch_cid.to_bytes())?;
+    //         }
+    //     } else {
+    //         // if no majority store vote and return
+    //         votes_map.set(ch_cid.to_bytes().into(), votes)?;
+    //     }
+    //
+    //     // flush votes
+    //     st.window_checks = votes_map.flush()?;
+    //
+    //     st.save();
+    //     Ok(None)
+    // }
+}
 
-        // get stake to know how much to release
-        let bt = make_map_with_root::<_, BigIntDe>(&st.stake, &Blockstore)?;
-        let stake = get_stake(&bt, &caller.clone())?;
-        if stake == TokenAmount::zero() {
-            abort!(USR_ILLEGAL_STATE, "caller has no stake in subnet");
+impl ActorCode for Actor {
+    fn invoke_method<BS, RT>(
+        rt: &mut RT,
+        method: MethodNum,
+        params: &RawBytes,
+    ) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        match FromPrimitive::from_u64(method) {
+            Some(Method::Constructor) => {
+                Self::constructor(rt, cbor::deserialize_params(params)?)?;
+                Ok(RawBytes::default())
+            }
+            Some(Method::Join) => {
+                let res = Self::join(rt, cbor::deserialize_params(params)?)?;
+                Ok(RawBytes::serialize(res)?)
+            }
+            None => Err(actor_error!(unhandled_message; "Invalid method")),
         }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Can we move these tests to its own file as we had before? That way we don't keep all the code in the same place.